File: Golang problems/Medium/1319_Number of Operations to Make Network Connected.go
/*
There are n computers numbered from 0 to n - 1 connected by ethernet cables connections forming a network where connections[i] = [ai, bi] represents a connection between computers ai and bi. Any computer can reach any other computer directly or indirectly through the network.

You are given an initial computer network connections. You can extract certain cables between two directly connected computers, and place them between any pair of disconnected computers to make them directly connected.

Return the minimum number of times you need to do this in order to make all the computers connected. If it is not possible, return -1.



Example 1:


Input: n = 4, connections = [[0,1],[0,2],[1,2]]
Output: 1
Explanation: Remove cable between computer 1 and 2 and place between computers 1 and 3.
Example 2:


Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2],[1,3]]
Output: 2
Example 3:

Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2]]
Output: -1
Explanation: There are not enough cables.


Constraints:

1 <= n <= 105
1 <= connections.length <= min(n * (n - 1) / 2, 105)
connections[i].length == 2
0 <= ai, bi < n
ai != bi
There are no repeated connections.
No two computers are connected by more than one cable.
*/

package main

import "fmt"

func makeConnected(n int, connections [][]int) int {
	// Создаем граф с помощью списка смежности
	graph := make(map[int][]int)
	for _, conn := range connections {
		graph[conn[0]] = append(graph[conn[0]], conn[1])
		graph[conn[1]] = append(graph[conn[1]], conn[0])
	}

	// Выполняем поиск в глубину, начиная с первой вершины графа
	visited := make(map[int]bool)
	dfs(graph, visited, 0)

	// Подсчитываем количество компонент связности графа
	numComponents := 1
	for i := 1; i < n; i++ {
		if !visited[i] {
			numComponents++
			dfs(graph, visited, i)
		}
	}

	// Подсчитываем количество ребер, которые нужно удалить
	numEdges := len(connections)
	if numEdges >= n-1 {
		return numComponents - 1
	}
	return -1
}

func dfs(graph map[int][]int, visited map[int]bool, node int) {
	visited[node] = true
	for _, neighbor := range graph[node] {
		if !visited[neighbor] {
			dfs(graph, visited, neighbor)
		}
	}
}

/*
Данная задача требует решения с помощью алгоритма поиска в глубину (DFS).
Мы можем представить данную сеть компьютеров как граф, где каждый компьютер представляет вершину, а соединения между компьютерами - это ребра графа.
Если мы можем достичь каждого компьютера от каждого другого компьютера, то граф называется связным.

Таким образом, задача сводится к поиску минимального количества ребер, которые нужно удалить из графа, чтобы сделать его связным.
Если граф уже связный, то мы не должны удалять никаких ребер.

Алгоритм:
Создать граф с помощью списка смежности на основе входных данных.
Выполнить поиск в глубину, начиная с первой вершины графа. Если мы можем достичь всех вершин графа из первой вершины, то граф уже связный, и мы не должны удалять никаких ребер.
Если граф не связный, то мы должны найти все компоненты связности графа и подсчитать количество ребер, которые нужно удалить, чтобы сделать каждую компоненту связной.
Возвращаем общее количество ребер, которые нужно удалить, чтобы сделать граф связным.
*/

func main() {
	fmt.Println(makeConnected(4, [][]int{{0, 1}, {0, 2}, {1, 2}}))
	fmt.Println(makeConnected(6, [][]int{{0, 1}, {0, 2}, {0, 3}, {1, 2}, {1, 3}}))
	fmt.Println(makeConnected(6, [][]int{{0, 1}, {0, 2}, {0, 3}, {1, 2}}))

}

File: Python problems/Medium/1319_Number of Operations to Make Network Connected.py
# There are n computers numbered from 0 to n - 1 connected by ethernet cables connections forming a network where connections[i] = [ai, bi] represents a connection between computers ai and bi. Any computer can reach any other computer directly or indirectly through the network.
#
# You are given an initial computer network connections. You can extract certain cables between two directly connected computers, and place them between any pair of disconnected computers to make them directly connected.
#
# Return the minimum number of times you need to do this in order to make all the computers connected. If it is not possible, return -1.
#
#
#
# Example 1:
#
#
# Input: n = 4, connections = [[0,1],[0,2],[1,2]]
# Output: 1
# Explanation: Remove cable between computer 1 and 2 and place between computers 1 and 3.
# Example 2:
#
#
# Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2],[1,3]]
# Output: 2
# Example 3:
#
# Input: n = 6, connections = [[0,1],[0,2],[0,3],[1,2]]
# Output: -1
# Explanation: There are not enough cables.
#
#
# Constraints:
#
# 1 <= n <= 105
# 1 <= connections.length <= min(n * (n - 1) / 2, 105)
# connections[i].length == 2
# 0 <= ai, bi < n
# ai != bi
# There are no repeated connections.
# No two computers are connected by more than one cable.


from collections import deque


class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1

        # create adjacency list
        adj_list = [[] for _ in range(n)]
        for u, v in connections:
            adj_list[u].append(v)
            adj_list[v].append(u)

        visited = set()
        components = 0

        for i in range(n):
            if i not in visited:
                components += 1
                visited.add(i)
                queue = deque([i])

                while queue:
                    node = queue.popleft()
                    for neighbor in adj_list[node]:
                        if neighbor not in visited:
                            visited.add(neighbor)
                            queue.append(neighbor)

        return components - 1

# Для решения этой задачи можно использовать алгоритм обхода графа в глубину (DFS) или алгоритм обхода графа в ширину
# (BFS) для нахождения связных компонент графа.
#
# Мы начнем с одной вершины и будем искать все вершины, связанные с ней. Затем мы перейдем к следующей вершине,
# которая еще не была посещена, и найдем все вершины, связанные с ней, и так далее, пока мы не исследуем все вершины
# графа.
#
# Если граф уже полностью связан, то мы можем вернуть 0. В противном случае мы должны добавить ребро между двумя
# компонентами графа, которые не соединены, и продолжать до тех пор, пока все вершины не будут связаны.
#
# Мы будем использовать алгоритм обхода графа в ширину (BFS) для этого. Мы начнем с одной вершины и будем искать все
# вершины, связанные с ней. Затем мы добавим все несвязанные с этой вершиной ребра в очередь, которую мы будем
# использовать для хранения несвязанных компонентов графа.
#
# Когда мы исследуем все вершины, связанные с текущей, мы извлечем первый элемент из очереди и продолжим процесс,
# пока все вершины не будут связаны.
#
# Для определения, связаны ли две вершины графа, мы можем использовать алгоритм обхода графа в глубину (DFS) или
# алгоритм обхода графа в ширину (BFS).
#
# Асимптотическая сложность времени этого алгоритма составляет O(n log n), где n - количество вершин в графе.


if __name__ == '__main__':
    solution = Solution()
    print(solution.makeConnected(4, [[0, 1], [0, 2], [1, 2]]))
    print(solution.makeConnected(6, [[0, 1], [0, 2], [0, 3], [1, 2], [1, 3]]))
    print(solution.makeConnected(6, [[0, 1], [0, 2], [0, 3], [1, 2]]))
